packages/runtime/lifecycle/src/instance/overRestAPI/index.ts

```diff
@@ -32,43 +38,70 @@ import * as Rest from "@marlowe.io/runtime-rest-client";
-    create: (payload: CreateRequest) =>
-      Command.create(restAPI)(walletAPI)(payload),
-    applyInputs: (contractId: ContractId) => (provideInput: ProvideInput) =>
-      pipe(
-        restAPI.contracts.contract.get(contractId),
-        TE.chain((header) =>
-          TE.fromTask(getParties(walletAPI)(header.roleTokenMintingPolicyId))
-        ),
-        TE.chain((parties) =>
-          restAPI.contracts.contract.next(contractId)(
-            mkEnvironment(pipe(Date.now(), (date) => subMinutes(date, 15)))(
-              pipe(Date.now(), (date) => addMinutes(date, 15))
-            )
-          )(parties)
-        ),
-        TE.chain((next) =>
-          Command.applyInputs(restAPI)(walletAPI)(contractId)(
-            provideInput(next)
-          )
+class ContractLifecycle implements ContractsAPI {
+  constructor(private wallet: WalletAPI, private rest: RestAPI) {}
+  async create(req: CreateRequest): Promise<ContractId> {
+    return unsafeTaskEither(Command.create(this.rest)(this.wallet)(req));
+  }
+
+  async applyInputs(
+    contractId: ContractId,
+    provideInput: ProvideInput
+  ): Promise<ContractId> {
+    const contractDetails = await unsafeTaskEither(
+      this.rest.contracts.contract.get(contractId)
+    );
+    const parties = await getParties(this.wallet)(
+      contractDetails.roleTokenMintingPolicyId
+    )();
+    const next = await unsafeTaskEither(
+      this.rest.contracts.contract.next(contractId)(
+        mkEnvironment(pipe(Date.now(), (date) => subMinutes(date, 15)))(
+          pipe(Date.now(), (date) => addMinutes(date, 15))
         )
-      ),
-  },
-  payouts: {
-    available: (filtersOption: O.Option<Filters>) =>
-      availablePayouts(restAPI)(walletAPI)(filtersOption),
-    withdraw: (payoutIds: PayoutId[]) =>
-      Command.withdraw(restAPI)(walletAPI)(payoutIds),
-    withdrawn: (filtersOption: O.Option<Filters>) =>
-      withdrawnPayouts(restAPI)(walletAPI)(filtersOption),
-  },
-});
+      )(parties)
+    );
+
+    return unsafeTaskEither(
+      Command.applyInputs(this.rest)(this.wallet)(contractId)(
+        provideInput(next)
+      )
+    );
+  }
+}
+
+class PayoutLifecycle implements PayoutsAPI {
+  constructor(private wallet: WalletAPI, private rest: RestAPI) {}
+
+  async available(filters?: Filters): Promise<PayoutAvailable[]> {
+    return unsafeTaskEither(
+      availablePayouts(this.rest)(this.wallet)(O.fromNullable(filters))
+    );
+  }
+
+  async withdraw(payoutIds: PayoutId[]): Promise<void> {
+    return unsafeTaskEither(
+      Command.withdraw(this.rest)(this.wallet)(payoutIds)
+    );
+  }
+
+  async withdrawn(filters?: Filters): Promise<PayoutWithdrawn[]> {
+    return unsafeTaskEither(
+      withdrawnPayouts(this.rest)(this.wallet)(O.fromNullable(filters))
+    );
+  }
+}
 
+export function mkRuntimeLifecycle(
+  restAPI: RestAPI,
+  wallet: WalletAPI
+): RuntimeLifecycle {
+  return {
+    wallet: wallet,
+    contracts: new ContractLifecycle(wallet, restAPI),
```

Comment:
not obvious, it's better using classes rather than directing implementing like it was before... When you use `new` somehow you open the door for a state... I don't know, I need to digest that...

Reply:
Yeah, I'm not married to it, as I mentioned @bjornkihlberg I still think that using a DI pattern with a two-level function could be useful. I wouldn't do proper currying because we don't gain much and it just looks weird in JS to be doing `foo(a1) (a2)(...)`

```typescript
type WalletDI = { wallet: WalletAPI }
type RuntimeRestDI = { rest: RestAPI}

type FooDI = WalletDI & RuntimeRestDI;

const foo = (di: FooDI) => (foo1: string, foo2: number) => {
  di.wallet.xxx();
}
```

To help you ease the `door for state`, think that the `wallet` for example, is a mutable object.

Reply:
I have give up on DI 10 years ago lol... I prefer composition...

Reply:
To be clear; `new` doesn't open any doors to state. That door is wide open whatever you do.